conformance/utils/roundtripper/roundtripper.go

```diff
@@ -0,0 +1,160 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package roundtripper
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"regexp"
+	"time"
+
+	"sigs.k8s.io/gateway-api/apis/v1alpha2"
+)
+
+// RoundTripper is an interface used to make requests within conformance tests.
+// This can be overridden with custom implementations whenever necessary.
+type RoundTripper interface {
+	CaptureRoundTrip(Request) (*CapturedRequest, *CapturedResponse, error)
+}
```

Comment:
These are great questions, I struggled with the appropriate interface here, but ended up with something that was pretty similar to what [ingress-controller-conformance implemented](https://github.com/kubernetes-sigs/ingress-controller-conformance/blob/master/test/http/http.go#L39-L66). I think there are 3 main reasons I avoided the net/http interfaces:

1. I want to be able to support more protocols than HTTP. I actually don't know how to test UDP, but TCP could theoretically recycle the same interface.
2. I want tests to be able to specify a request to example.com/foo and know that it will get routed to the appropriate Gateway automatically. That could technically be done by modifying/copying a http request that was passed in, but feels messier. We could also just make each test case specify this as they structure a request.
3. I wanted to use the same echoserver image as ingress-controller-conformance so we didn't have to maintain a separate one for this. That requires a struct that matches the output from that container.

Open to different approaches here, but that's how I landed up with the current approach.